pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,522 @@
+	schema.GroupVersionResource{Group: "authorization.k8s.io", Version: "v1beta1", Resource: "localsubjectaccessreviews"}: {},
+}
+
+func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynamic.ClientPool, mapper meta.RESTMapper, deletableResources map[schema.GroupVersionResource]struct{}) (*GarbageCollector, error) {
+	attemptToDelete := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_delete")
+	attemptToOrphan := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_orphan")
+	absentOwnerCache := NewUIDCache(500)
+	gc := &GarbageCollector{
+		metaOnlyClientPool:                  metaOnlyClientPool,
+		clientPool:                          clientPool,
+		restMapper:                          mapper,
+		attemptToDelete:                     attemptToDelete,
+		attemptToOrphan:                     attemptToOrphan,
+		registeredRateLimiter:               NewRegisteredRateLimiter(deletableResources),
+		registeredRateLimiterForControllers: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:                    absentOwnerCache,
+	}
+	gc.dependencyGraphBuilder = &GraphBuilder{
+		graphChanges: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_graph_changes"),
+		uidToNode: &concurrentUIDToNode{
+			RWMutex:   &sync.RWMutex{},
+			uidToNode: make(map[types.UID]*node),
+		},
+		attemptToDelete:  attemptToDelete,
+		attemptToOrphan:  attemptToOrphan,
+		absentOwnerCache: absentOwnerCache,
+	}
+	for resource := range deletableResources {
+		if _, ok := ignoredResources[resource]; ok {
+			glog.V(5).Infof("ignore resource %#v", resource)
+			continue
+		}
+		kind, err := gc.restMapper.KindFor(resource)
+		if err != nil {
+			return nil, err
+		}
+		controller, err := gc.controllerFor(resource, kind)
+		if err != nil {
+			return nil, err
+		}
+		gc.monitors = append(gc.monitors, monitor)
+	}
+	return gc, nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.attemptToDelete.ShutDown()
+	defer gc.attemptToOrphan.ShutDown()
+	defer gc.dependencyGraphBuilder.graphChanges.ShutDown()
+
+	glog.Infof("Garbage Collector: Initializing")
+	for _, monitor := range gc.monitors {
+		go monitor.Run(stopCh)
+	}
+
+	var syncs []cache.InformerSynced
+	for _, monitor := range gc.monitors {
+		syncs = syncs.append(monitor.HasSynced())
+	}
+	if !cache.WaitForCacheSync(stopCh, syncs...) {
+		return
+	}
+	glog.Infof("Garbage Collector: All monitored resources synced. Proceeding to collect garbage")
+
+	// worker
+	go wait.Until(gc.dependencyGraphBuilder.processEvent, 0, stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.attemptToDeleteWorker, 0, stopCh)
+		go wait.Until(gc.attemptToOrphanWorker, 0, stopCh)
+	}
+	Register()
+	<-stopCh
+	glog.Infof("Garbage Collector: Shutting down")
+}
+
+func (gc *GarbageCollector) attemptToDeleteWorker() {
```

Comment:
> Actually I think the structure of the linked controller is bloated. wait.Until already has a infinite for loop, the controller doesn't need to define another runWorker() function to run the business logic infinitely.

Those loops do different things.  The `wait.Until` will restart on failure and one failures will prevent hotlooping since it has a logical wait.  The `runWorker` hotloops when busy, blocks until work comes in, but has no backoff on failures.